opensearch/src/main/java/org/opensearch/sql/opensearch/storage/scan/BackgroundSearchScanner.java

```diff
@@ -0,0 +1,177 @@
+  @Nullable private final Executor backgroundExecutor;
+  private CompletableFuture<OpenSearchResponse> nextBatchFuture = null;
+  private boolean stopIteration = false;
+
+  public BackgroundSearchScanner(OpenSearchClient client) {
+    this.client = client;
+    // We can only actually do the background operation if we have the ability to access the thread
+    // pool. Otherwise, fallback to synchronous fetch.
+    if (client.getNodeClient().isPresent()) {
+      this.backgroundExecutor =
+          client.getNodeClient().get().threadPool().executor(SQL_BACKGROUND_THREAD_POOL_NAME);
+    } else {
+      this.backgroundExecutor = null;
+    }
+  }
+
+  private boolean isAsync() {
+    return backgroundExecutor != null;
+  }
+
+  /**
+   * @return Whether the search scanner has fetched all batches
+   */
+  public boolean isScanDone() {
+    return stopIteration;
+  }
+
+  /**
+   * Initiates the scanning process. If async operations are available, this will trigger the first
+   * background fetch.
+   *
+   * @param request The OpenSearch request to execute
+   */
+  public void startScanning(OpenSearchRequest request) {
+    if (isAsync()) {
+      nextBatchFuture =
+          CompletableFuture.supplyAsync(() -> client.search(request), backgroundExecutor);
+    }
+  }
+
+  private OpenSearchResponse getCurrentResponse(OpenSearchRequest request) {
+    if (isAsync()) {
+      try {
+        return nextBatchFuture.get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new NonFallbackCalciteException(
+            "Failed to fetch data from the index: the background task failed or interrupted.\n"
+                + "  Inner error: "
+                + e.getMessage());
+      }
+    } else {
+      return client.search(request);
+    }
+  }
+
+  /**
+   * Fetches the next batch of results. If async is enabled and more batches are expected, this will
+   * also trigger the next background fetch.
+   *
+   * @param request The OpenSearch request to execute
+   * @param maxResultWindow Maximum number of results to fetch per batch
+   * @return SearchBatchResult containing the current batch's iterator and completion status
+   * @throws NonFallbackCalciteException if the background fetch fails or is interrupted
+   */
+  public SearchBatchResult fetchNextBatch(OpenSearchRequest request, int maxResultWindow) {
+    OpenSearchResponse response = getCurrentResponse(request);
+
+    // Determine if we need future batches
+    if (response.isAggregationResponse()
+        || response.isCountResponse()
+        || response.getHitsSize() < maxResultWindow) {
+      stopIteration = true;
+    }
+
+    Iterator<ExprValue> iterator;
+    if (!response.isEmpty()) {
+      iterator = response.iterator();
+
+      // Pre-fetch next batch if needed
+      if (!stopIteration && isAsync()) {
+        nextBatchFuture =
+            CompletableFuture.supplyAsync(() -> client.search(request), backgroundExecutor);
```

Comment:
what if fixedThreadPool full, should fallback to sync?

Reply:
We just buffer: https://github.com/opensearch-project/sql/pull/4345#discussion_r2383251023